add metrics framework

pwinckles opened the pull request:
**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3443

# What does this Pull Request do?

This PR is intended to demonstration a possible metrics integration. It uses [Micrometer](https://micrometer.io/) for collecting metrics within the app and reports them to [Prometheus](https://prometheus.io/). Micrometer could also be configured to report to any number of alternative metrics backends. Prometheus is very powerful, though it is not the most intuitive to use.

Regardless, if we do want to support capturing metrics, I think it would likely make sense to offer a number of different backends so that the metrics could more easily integrate with whatever service an institution is already using. For example, [Graphite](https://graphiteapp.org/) is also popular. However, the bigger question is in regards to if we want to use Micrometer or an alternative like [Dropwizard](https://metrics.dropwizard.io/4.1.2/). Micrometer is more popular currently and perhaps better designed, but Dropwizard does have some out-of-the-box [Jersey integrations](https://metrics.dropwizard.io/4.1.2/manual/jersey.html) that Micrometer does not have.

A word about annotations, out of the box, Micrometer only supports AspectJ for annotation based annotation based metrics collection. I do not think this is a bad thing. Spring proxy based AOP is not a good fit for annotation based metrics collection because it has serious limitations which would likely result in people adding annotations that they think would emit metrics when they are actually doing nothing.

Prometheus works by periodically "scraping" metrics. I added a new endpoint, `/prometheus` to the webapp that exposes the metrics for Prometheus. I exposed some default JVM metrics and added two timing metrics on get and put resource requests. No other custom metrics have been added at this point.

# How should this be tested?

Write the following to a file named `prometheus.yml`:

```yml
scrape_configs:
  - job_name: 'fcrepo'
    scrape_interval: 30s 
    metrics_path: '/prometheus'
    static_configs:
    - targets: ['host.docker.internal:8080']
    basic_auth:
      username: fedoraAdmin
      password: fedoraAdmin
```

Write the following to a file named `docker-compose.yml` that is in the same directory as `prometheus.yml`:

```yml
version: '3.8'
services:
 prometheus:
   image: prom/prometheus
   ports:
     - "9090:9090"
   volumes:
     - ./prometheus.yml:/etc/prometheus/prometheus.yml
    
 grafana:
   image: grafana/grafana
   ports:
   - "3000:3000"
```

Execute: `docker-compose up`

Finally, startup Fedora with metrics enabled (`-Dfcrepo.metrics.enable=true`) and execute a few requests like:

```
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/binary -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "testing"
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/binary
```

Navigate to http://localhost:9090, to query the metrics directly in Prometheus or go to http://localhost:3000, if you'd like to use Grafana.

If you're using Grafana, here are the steps to connect it to Prometheus:

1. Login: admin/admin
1. Configuration -> Datasources
1. Add datasource -> Prometheus
1. URL: http://prometheus:9090
1. Save & Test
1. Create a new dashboard

**[EDIT]** The following queries don't work anymore because I removed the metrics. Load [this dashboard](https://github.com/fcrepo4/fcrepo4/files/5285946/grafana.txt) into Grafana for some example queries.

Here are some example queries:

* JVM memory usage: `jvm_memory_used_bytes`
* All API requests throughput: `rate(fcrepo_http_request_seconds_count[1m])`
* GetResource API request throughput: `rate(fcrepo_http_request_seconds_count{api="getResource"}[1m])`
* GetResource API mean request time: `rate(fcrepo_http_request_seconds_sum{api="getResource"}[1m])/rate(fcrepo_http_request_seconds_count{api="getResource"}[1m])`
* GetResource API p99 request time: `histogram_quantile(0.99, sum(rate(fcrepo_http_request_seconds_bucket{api="getResource"}[1m])) by (le))`

# Interested parties
@fcrepo4/committers @awoods

whikloj commented:
🤣 I present to you, `fcrepo-metrics` https://github.com/fcrepo4/fcrepo4/tree/4.7-maintenance/fcrepo-metrics

pwinckles commented:
> 
> 
> 🤣 I present to you, `fcrepo-metrics` https://github.com/fcrepo4/fcrepo4/tree/4.7-maintenance/fcrepo-metrics

What is the story there?

pwinckles commented:
Actually, Micrometer does have Jersey 2 integration. Let me see if I can get it hooked up

whikloj commented:
We used to have these test bits in the codebase to allow for metrics much like (I assume) what you have done here. But no one was using them and the setup with Graphite was a lot of work. So it got removed. https://jira.lyrasis.org/browse/FCREPO-2779

pwinckles commented:
> 
> 
> We used to have these test bits in the codebase to allow for metrics much like (I assume) what you have done here. But no one was using them and the setup with Graphite was a lot of work. So it got removed. https://jira.lyrasis.org/browse/FCREPO-2779

I see, that I would have been useful context to have known earlier.

whikloj commented:
If you can make the bar a little lower I'd be interested. I had tried to use Graphite before, but it wasn't super friendly and I got distracted.

awoods commented:
Here is some old documentation: https://wiki.lyrasis.org/display/FEDORA4x/Setup+a+Graphite+instance

awoods commented on fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/PrometheusMetricsServlet.java:
```diff
@@ -0,0 +1,52 @@
+        try {
+            final var field = MetricsServlet.class.getDeclaredField("registry");
+            field.setAccessible(true);
+            field.set(this, collector);
```
Ouch. No injection possibilities?! :(

pwinckles replied:
I'm open to ideas. I don't know of a way of wiring a servlet's constructor because it needs to be instantiated by the container.

awoods reviewed (commented):
This is an excellent starting point. I would like to brainstorm effective use of the Grafana dashboard and common queries. 
Additionally, it appears that the `/prometheus` endpoint is not available with the one-click.

pwinckles commented:
@awoods I fixed the one-click (didn't realize it had its own web.xml) and added the micrometer jersey integration. Now, all of our Jersey APIs are reporting metrics. I removed the manual metrics that I had added, so the queries in the PR description are no longer good.

Here is an example Grafana dashboard that has a few queries setup:

```json
{
  "__inputs": [
    {
      "name": "DS_PROMETHEUS",
      "label": "Prometheus",
      "description": "",
      "type": "datasource",
      "pluginId": "prometheus",
      "pluginName": "Prometheus"
    }
  ],
  "__requires": [
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "7.2.0"
    },
    {
      "type": "panel",
      "id": "graph",
      "name": "Graph",
      "version": ""
    },
    {
      "type": "datasource",
      "id": "prometheus",
      "name": "Prometheus",
      "version": "1.0.0"
    }
  ],
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": null,
  "links": [],
  "panels": [
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "${DS_PROMETHEUS}",
      "fieldConfig": {
        "defaults": {
          "custom": {}
        },
        "overrides": []
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 4,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "7.2.0",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "expr": "rate(http_server_requests_seconds_count[1m])",
          "interval": "",
          "legendFormat": "",
          "refId": "A"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "Request Rate",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    },
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "${DS_PROMETHEUS}",
      "fieldConfig": {
        "defaults": {
          "custom": {}
        },
        "overrides": []
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 2,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "7.2.0",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "expr": "histogram_quantile(0.99, sum(rate(http_server_requests_seconds_bucket{method=\"GET\",uri=\"/rest/{path: .*}\"}[1m])) by (le))",
          "interval": "",
          "legendFormat": "p99",
          "refId": "A"
        },
        {
          "expr": "rate(http_server_requests_seconds_sum{method=\"GET\",uri=\"/rest/{path: .*}\"}[1m])/rate(http_server_requests_seconds_count{method=\"GET\",uri=\"/rest/{path: .*}\"}[1m])",
          "interval": "",
          "legendFormat": "avg",
          "refId": "B"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "GET Resource Time",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    },
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "${DS_PROMETHEUS}",
      "fieldConfig": {
        "defaults": {
          "custom": {}
        },
        "overrides": []
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 8
      },
      "hiddenSeries": false,
      "id": 8,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "7.2.0",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "expr": "rate(http_server_requests_seconds_count{outcome=~\".*_ERROR\"}[1m])",
          "interval": "",
          "legendFormat": "",
          "refId": "A"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "Request Error Rate",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    },
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "${DS_PROMETHEUS}",
      "fieldConfig": {
        "defaults": {
          "custom": {}
        },
        "overrides": []
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 12,
        "y": 9
      },
      "hiddenSeries": false,
      "id": 6,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "7.2.0",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "expr": "jvm_memory_used_bytes",
          "interval": "",
          "legendFormat": "",
          "refId": "A"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "JVM Memory Usage",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    }
  ],
  "schemaVersion": 26,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-1h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Example Fedora Dashboard",
  "uid": "AEUsf4KMk",
  "version": 1
}
```

whikloj commented on fcrepo-configs/src/main/java/org/fcrepo/config/MetricsConfig.java:
```diff
@@ -0,0 +1,101 @@
+@Configuration
+public class MetricsConfig {
+
+    @Value("${fcrepo.metrics.enable:true}")
```
Wondering if the default should be `false`, will most users care about this?

pwinckles replied:
Yes, I think that is reasonable

whikloj replied:
Also, what is the impact (if any) by having this running but not actually using the metrics?

pwinckles replied:
There should not be a difference between collecting the metrics and not using them and collecting the metrics and using them.

There is a difference between collecting the metrics and not collecting the metrics, but I cannot approximate the memory and cpu cost.

awoods commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java:
```diff
@@ -17,113 +17,112 @@
  * @since 9/25/14
  */
 
+@Timed
```
There are a lot of changes to this file just to add a single annotation: `@Timed`.
I would suggest only reformatting code that is being directly changed.

awoods reviewed (changes requested):
This should be good to go... pending response to one comment.

awoods commented:
This is actually a significant feature for Fedora... worthy of a JIRA ticket.

pwinckles commented:
> 
> 
> This is actually a significant feature for Fedora... worthy of a JIRA ticket.

https://jira.lyrasis.org/browse/FCREPO-3443
